File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/ui/controller/StyledToolbarNavigationController.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.adamantcheese.chan.ui.controller;

import android.content.Context;

import com.github.adamantcheese.chan.R;
import com.github.adamantcheese.chan.controller.Controller;
import com.github.adamantcheese.chan.controller.ControllerTransition;
import com.github.adamantcheese.chan.controller.ui.NavigationControllerContainerLayout;
import com.github.adamantcheese.chan.core.settings.ChanSettings;
import com.github.adamantcheese.chan.features.drawer.DrawerController;
import com.github.adamantcheese.chan.ui.theme.ThemeHelper;

import javax.inject.Inject;

import static com.github.adamantcheese.chan.Chan.inject;
import static com.github.adamantcheese.chan.utils.AndroidUtils.inflate;

public class StyledToolbarNavigationController extends ToolbarNavigationController {

    @Inject
    ThemeHelper themeHelper;

    public StyledToolbarNavigationController(Context context) {
        super(context);
        inject(this);
    }

    @Override
    public void onCreate() {
        super.onCreate();

        view = inflate(context, R.layout.controller_navigation_toolbar);
        container = (NavigationControllerContainerLayout) view.findViewById(R.id.container);

        NavigationControllerContainerLayout nav = (NavigationControllerContainerLayout) container;
        nav.setNavigationController(this);
        nav.setSwipeEnabled(ChanSettings.controllerSwipeable.get());

        toolbar = view.findViewById(R.id.toolbar);
        toolbar.setBackgroundColor(themeHelper.getTheme().primaryColor.color);
        toolbar.setCallback(this);
    }

    @Override
    public boolean popController(ControllerTransition controllerTransition) {
        return !toolbar.isTransitioning() && super.popController(controllerTransition);
    }

    @Override
    public boolean pushController(Controller to, ControllerTransition controllerTransition) {
        return !toolbar.isTransitioning() && super.pushController(to, controllerTransition);
    }

    @Override
    public void transition(Controller from, Controller to, boolean pushing, ControllerTransition controllerTransition) {
        super.transition(from, to, pushing, controllerTransition);

        if (to != null) {
            DrawerController drawerController = getDrawerController();
            if (drawerController != null) {
                drawerController.setDrawerEnabled(to.navigation.hasDrawer);
            }
        }
    }

    @Override
    public void endSwipeTransition(Controller from, Controller to, boolean finish) {
        super.endSwipeTransition(from, to, finish);

        if (finish) {
            DrawerController drawerController = getDrawerController();
            if (drawerController != null) {
                drawerController.setDrawerEnabled(to.navigation.hasDrawer);
            }
        }
    }

    @Override
    public boolean onBack() {
        if (super.onBack()) {
            return true;
        } else if (parentController instanceof PopupController && childControllers.size() == 1) {
            ((PopupController) parentController).dismiss();
            return true;
        } else if (doubleNavigationController != null && childControllers.size() == 1) {
            if (doubleNavigationController.getRightController() == this) {
                doubleNavigationController.setRightController(null);
                return true;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    @Override
    public void onMenuClicked() {
        DrawerController drawerController = getDrawerController();
        if (drawerController != null) {
            drawerController.onMenuClicked();
        }
    }

    private DrawerController getDrawerController() {
        if (parentController instanceof DrawerController) {
            return (DrawerController) parentController;
        } else if (doubleNavigationController != null) {
            Controller doubleNav = (Controller) doubleNavigationController;
            if (doubleNav.parentController instanceof DrawerController) {
                return (DrawerController) doubleNav.parentController;
            }
        }
        return null;
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/navigation/StyledToolbarNavigationController.kt
package com.github.k1rakishou.chan.ui.controller.navigation

import android.content.Context
import android.view.ViewGroup
import com.github.k1rakishou.ChanSettings
import com.github.k1rakishou.chan.R
import com.github.k1rakishou.chan.controller.Controller
import com.github.k1rakishou.chan.controller.transition.ControllerTransition
import com.github.k1rakishou.chan.controller.ui.NavigationControllerContainerLayout
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent
import com.github.k1rakishou.chan.features.drawer.MainController
import com.github.k1rakishou.chan.features.toolbar_v2.KurobaToolbarView
import com.github.k1rakishou.chan.ui.controller.PopupController
import com.github.k1rakishou.chan.ui.controller.ThreadSlideController
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils

class StyledToolbarNavigationController(context: Context) : ToolbarNavigationController(context) {

  private val mainController: MainController?
    get() {
      if (parentController is MainController) {
        return parentController as MainController?
      } else if (doubleNavigationController != null) {
        val doubleNav = doubleNavigationController as Controller
        if (doubleNav.parentController is MainController) {
          return doubleNav.parentController as MainController?
        }
      }

      return null
    }

  override fun injectDependencies(component: ActivityComponent) {
    component.inject(this)
  }

  override fun onCreate() {
    super.onCreate()

    view = AppModuleAndroidUtils.inflate(context, R.layout.controller_navigation_toolbar)
    container = view.findViewById<ViewGroup>(R.id.toolbar_navigation_controller_container)

    val toolbar = view.findViewById<KurobaToolbarView>(R.id.toolbar)
    toolbar.init(this)
    toolbarState.enterContainerMode()

    reloadControllerTracking()
  }

  override fun popController(transition: ControllerTransition?): Boolean {
    val result = super.popController(transition)
    reloadControllerTracking()
    return result
  }

  override fun pushController(to: Controller, transition: ControllerTransition?): Boolean {
    val result = super.pushController(to, transition)
    reloadControllerTracking()
    return result
  }

  fun onChildControllerPushed(controller: Controller) {
    reloadControllerTracking()
  }

  fun onChildControllerPopped(controller: Controller) {
    reloadControllerTracking()
  }

  private fun reloadControllerTracking() {
    if (container == null) {
      return
    }

    val nav = container as NavigationControllerContainerLayout
    val threadSlideController = threadSlideControllerOrNull()

    if (topController == null || threadSlideController != null) {
      val viewThreadController = threadSlideController?.rightController()
      if (viewThreadController != null) {
        if (ChanSettings.viewThreadControllerSwipeable.get()) {
          nav.initThreadControllerTracking(this)
        } else {
          nav.initThreadDrawerOpenGestureControllerTracker(this)
        }

        return
      }

      // fallthrough
    }

    if (ChanSettings.controllerSwipeable.get()) {
      nav.initThreadControllerTracking(this)
    } else {
      nav.initThreadDrawerOpenGestureControllerTracker(this)
    }
  }

  private fun threadSlideControllerOrNull(): ThreadSlideController? {
    val top = topController
    if (top is ThreadSlideController) {
      return top
    }

    return null
  }

  override fun transition(
    from: Controller?,
    to: Controller?,
    pushing: Boolean,
    controllerTransition: ControllerTransition?
  ) {
    if (from == null && to == null) {
      return
    }

    super.transition(from, to, pushing, controllerTransition)

    if (to != null) {
      val mainController = mainController
      mainController?.onNavigationItemDrawerInfoUpdated(to.hasDrawer)
    }
  }

  override fun endSwipeTransition(from: Controller?, to: Controller?, finish: Boolean) {
    if (from == null && to == null) {
      return
    }

    super.endSwipeTransition(from, to, finish)

    if (finish && to != null) {
      val mainController = mainController
      mainController?.onNavigationItemDrawerInfoUpdated(to.hasDrawer)
    }
  }

  override fun onBack(): Boolean {
    if (super.onBack()) {
      return true
    }

    if (parentController is PopupController && childControllers.size == 1) {
      (parentController as PopupController).dismiss()
      return true
    }

    if (doubleNavigationController != null && childControllers.size == 1) {
      if (doubleNavigationController!!.rightController() === this) {
        doubleNavigationController!!.updateRightController(null, false)
        return true
      }
      return false
    }

    return false
  }

}
